File: backend-python/app/routers/jobs.py
import uuid
import os
import shutil
from fastapi import APIRouter, HTTPException, Query, Depends, UploadFile, File
from fastapi.responses import JSONResponse
from sqlalchemy.orm import Session
from app.database import get_db
from app.models import Job
from app.schemas import JobCreate, JobResponse, JobListResponse
from app.tasks import dispatch

UPLOAD_DIR = os.environ.get("UPLOAD_DIR", "/uploads")
os.makedirs(UPLOAD_DIR, exist_ok=True)

router = APIRouter()


# POST /upload
@router.post("/upload")
def upload_file(file: UploadFile = File(...)):
    dest = os.path.join(UPLOAD_DIR, file.filename)
    with open(dest, "wb") as f:
        shutil.copyfileobj(file.file, f)
    return JSONResponse({"path": dest})


# POST /jobs
@router.post("/", response_model=JobResponse, status_code=201)
def submit_job(body: JobCreate, db: Session = Depends(get_db)):
    job = Job(
        type=body.type,
        payload=body.payload,
        max_retries=body.max_retries,
        status="pending",
    )
    db.add(job)
    db.commit()
    db.refresh(job)

    # Enqueue the job. If the broker is unreachable, don't leave the row
    # stranded as "pending" - mark it failed so its state is honest.
    try:
        dispatch(str(job.id), job.type, job.payload)
    except Exception as exc:
        job.status = "failed"
        job.error = f"dispatch_failed: {exc}"
        db.commit()
        db.refresh(job)
        raise HTTPException(
            status_code=503,
            detail="Job accepted but could not be queued; marked failed.",
        )

    return job


# GET /jobs
@router.get("/", response_model=JobListResponse)
def list_jobs(
    status: str | None = Query(None, description="Filter by job status"),
    limit: int = Query(20, le=100),
    offset: int = Query(0),
    db: Session = Depends(get_db),
):
    query = db.query(Job)
    if status:
        query = query.filter(Job.status == status)

    total = query.count()
    jobs = query.order_by(Job.created_at.desc()).offset(offset).limit(limit).all()

    return {"jobs": jobs, "total": total}


# GET /jobs/{id}
@router.get("/{job_id}", response_model=JobResponse)
def get_job(job_id: uuid.UUID, db: Session = Depends(get_db)):
    job = db.get(Job, job_id)
    if not job:
        raise HTTPException(status_code=404, detail="Job not found")
    return job


# DELETE /jobs/{id}
@router.delete("/{job_id}", status_code=204)
def delete_job(job_id: uuid.UUID, db: Session = Depends(get_db)):
    job = db.get(Job, job_id)
    if not job:
        raise HTTPException(status_code=404, detail="Job not found")
    db.delete(job)
    db.commit()

File: backend-node/src/routes/jobs.ts
import { Request, Router, Response } from "express";
import { v4 as uuidv4 } from "uuid";
import { query } from "../db";
import { dispatch, JobType } from "../queue";

const router = Router();

// POST /jobs - submit a new job
router.post("/", async (req: Request, res: Response): Promise<void> => {
  const {
    type,
    payload = {},
    max_retries = 3,
  } = req.body as {
    type: JobType;
    payload?: Record<string, unknown>;
    max_retries: number;
  };
  if (!type) {
    res.status(400).json({ error: "type is required" });
    return;
  }
  // Node backend is scrape-only; reject other types explicitly rather than
  // inserting a row no handler can process
  if (type !== "scrape") {
    res.status(400).json({
      error: `Unsupported job type '${type}' on the Node backend. Only 'scrape' is supported here; use the Python backend for resize/convert.`,
    });
    return;
  }

  const id = uuidv4();
  await query(
    `INSERT INTO jobs (id, type, payload, max_retries, status) VALUES ($1, $2, $3, $4, 'pending')`,
    [id, type, JSON.stringify(payload), max_retries],
  );

  // Enqueue the job. If the broker is unreachable, don't leave the row
  // stranded as "pending" - mark it failed so its state is honest.
  try {
    await dispatch(id, type, payload, max_retries);
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    await query(`UPDATE jobs SET status = 'failed', error = $2 WHERE id = $1`, [
      id,
      `dispatch_failed: ${message}`,
    ]);
    res
      .status(503)
      .json({ error: "Job accepted but could not be queued; marked failed." });
    return;
  }

  const [job] = await query(`SELECT * FROM jobs WHERE id = $1`, [id]);
  res.status(201).json(job);
});

// GET /jobs - list jobs
router.get("/", async (req: Request, res: Response): Promise<void> => {
  const {
    status,
    limit = "20",
    offset = "0",
  } = req.query as Record<string, string>;

  const conditions: string[] = [];
  const values: unknown[] = [];

  if (status) {
    conditions.push(`status=$${values.length + 1}`);
    values.push(status);
  }

  const where = conditions.length ? `WHERE ${conditions.join(" AND ")}` : "";
  const limitNum = Math.min(parseInt(limit), 100);
  const offsetNum = parseInt(offset);

  const jobs = await query(
    `SELECT * FROM jobs ${where} ORDER BY created_at DESC LIMIT $${values.length + 1} OFFSET $${values.length + 2}`,
    [...values, limitNum, offsetNum],
  );

  const [{ count }] = await query<{ count: string }>(
    `SELECT COUNT(*) FROM jobs ${where}`,
    values,
  );

  res.json({ jobs, total: parseInt(count) });
});

// GET /jobs/:id - get job
router.get("/:id", async (req: Request, res: Response): Promise<void> => {
  const [job] = await query(`SELECT * FROM jobs WHERE id = $1`, [
    req.params.id,
  ]);
  if (!job) {
    res.status(404).json({ error: "Job not found" });
    return;
  }
  res.json(job);
});

// DELETE /jobs/:id - remove job
router.delete("/:id", async (req: Request, res: Response): Promise<void> => {
  const [job] = await query(`SELECT * FROM jobs WHERE id = $1`, [
    req.params.id,
  ]);
  if (!job) {
    res.status(404).json({ error: "Job not found" });
    return;
  }
  await query(`DELETE FROM jobs WHERE id = $1`, [req.params.id]);
  res.status(204).send();
});

export default router;
